### PR TITLE
Chimera/build frameworks

### PR DESCRIPTION
**Pull request by laxmorek**
Working on #1674
Part of #2394

### Short description 📝

This PR adds logic for building XCFrameworks for SPM dependencies defined in `Dependencies.swift.`

**NOTE:** This PR targets to my WIP branch. It is not final version. My intention is to split implementation on smaller pieces and to collect feedback for ongoing work. More PRs are coming. 

### Implementation 👩‍💻👨‍💻

**The flow**
1. Load dependencies using `swift package show-dependencies`.
2. For `dependency` in  `dependencies` 
    1. Load `Package.swift` using `swift package dump-package`.
    2. Generate Xcode project using `swift package generate-xcodeproj`.
    3. Build frameworks for platforms defined in `Dependencies.swift` using `xcodebuild`.
    4. Combine frameworks into XCFrameworks using `xcodebuild`.

**Manifest**
 
`Dependencies.swift` structure has been changed since list of platforms is required by `Carthage` and `SPM`. Now `platforms` property is top level. 

```swift
import ProjectDescription

let dependencies = Dependencies(
    carthage: .carthage(
        [
            .github(path: "Alamofire/Alamofire", requirement: .exact("5.0.4"))
        ],
        options: [.useXCFrameworks, .noUseBinaries]
    ),
    swiftPackageManager: .swiftPackageManager(
        [
            .package(url: "https://github.com/ReactiveX/RxSwift.git", .exact("5.1.1"))
        ]
    ),
    platforms: [.iOS]
)
```

I can see an opportunity in adding `ExpressibleByArrayLiteral` to `CarthageDependencies` and `SwiftPackageManagerDependencies` to make creation of `Dependencies` more convenient. Im going to add it in one of the upcoming PRs. 

**Testing**

- `dependencies.feature` acceptance test has been updated. I have a plan to add more complex tests in one of the upcoming PRs. . 
- Im going to add integration tests iin one of the upcoming PRs.

### CODE REVIEW

File: Sources/TuistSupport/XcodeBuild/XcodeBuildController.swift [hunk 1]
```diff
@@ -1,8 +1,69 @@
...
 import TSCBasic
-import TuistCore
-import TuistSupport
+
+public enum XcodeBuildDestination: Equatable {
+    case device(String)
+    case mac
+}
+
+public protocol XcodeBuildControlling {
```
Comment (thedavidharris) [inline Sources/TuistSupport/XcodeBuild/XcodeBuildController.swift:10]: This isn't related to this PR, but I'm noticing that this isn't updated to include dSYMs and bcsymbolmaps in xcframeworks

**Review by thedavidharris — COMMENTED** [review]
Overall this looks good, just pointing out a few concerns I think I've mentioned in Slack as well:

`Generate Xcode project using swift package **generate-xcodeproj.`

The Swift Core team has noted they consider this functionality deprecated, both in the code at https://github.com/apple/swift-package-manager/pull/3062 and via discussion at tps://forums.swift.org/t/rfc-deprecating-generate-xcodeproj/42159. 

`Build frameworks for platforms defined in Dependencies.swift using xcodebuild.`

There are a few things here that won't necessarily work because of how some SPM-specific stuff works:

- Info.plist things can get wonky.  At least in the past `<string>$(CURRENT_PROJECT_VERSION)</string>` wasn't always present, which makes it an invalid bundle. This might be fixed, haven't checked in a while.
- Frameworks with resources won't work as the generate-xcodeproj doesn't include the generated code to access said resources, so you'll get `error: type 'Bundle' has no member 'module'` errors. That generated code looks pretty similar to how Tuist handles Bundle access:

```
import class Foundation.Bundle

private class BundleFinder {}

extension Foundation.Bundle {
    /// Returns the resource bundle associated with the current Swift module.
    static var module: Bundle = {
        let bundleName = "MyBundle"

        let candidates = [
            // Bundle should be present here when the package is linked into an App.
            Bundle.main.resourceURL,

            // Bundle should be present here when the package is linked into a framework.
            Bundle(for: BundleFinder.self).resourceURL,

            // For command-line tools.
            Bundle.main.bundleURL,
        ]

        for candidate in candidates {
            let bundlePath = candidate?.appendingPathComponent(bundleName + ".bundle")
            if let bundle = bundlePath.flatMap(Bundle.init(url:)) {
                return bundle
            }
        }
        fatalError("unable to find bundle named MyBundle")
    }()
}
```

**Comment by laxmorek** [comment]
@thedavidharris thank you for your review. 🙇‍♂️

> The Swift Core team has noted they consider this functionality deprecated

Yeah, that might be a problem in the future. We definitely don't want to build things on functionalities that will be removed. I'm going to check how `Carthage` team solved it. 

> This isn't related to this PR, but I'm noticing that this isn't updated to include dSYMs and bcsymbolmaps in xcframeworks

Yeah, I am going to add it later when we agree on the foundation of this feature. I didn't want to make this PR too big. 


> Build frameworks for platforms defined in Dependencies.swift using xcodebuild.
> There are a few things here that won't necessarily work because of how some SPM-specific stuff works:

Could we use `swift build` instead of `xcodebuild`? 🤔 Have you any experience with that?

**Comment by thedavidharris** [comment]
> Yeah, that might be a problem in the future. We definitely don't want to build things on functionalities that will be removed. I'm going to check how Carthage builds XCFrameworks

Carthage doesn't support SPM modules for this reason, you have to generate the xcodeproj somehow. It's partly why it's not really working as an ideal dependency manager anymore, as with SPM/CocoaPods you don't need it

**Comment by pepibumur** [comment]
👋 I think we need to take a step back here. Tuist shouldn't integrate those packages as .xcframeworks. Among other reasons, because of what @thedavidharris mentioned. The workflow should be the following:

1. We resolve the dependencies using the SPM.
2. Serialize the graph of third-party dependencies into a disk as a JSON.
3. Extend the current graph model to support references to nodes that are in the third-party dependencies graph.
4. Implement the logic to merge the 2 graphs at loading time.

To achieve that, I'd take the following steps in different PRs:
1. Make `ValueGraph` serializable (conform `Codable`)
2. Extend the `tuist dependencies update/fetch` to write the Graph to disk after resolving the dependencies.
3. Extend `ValueGraph` to add a new type of dependency:

```swift
// Here are some ideas for names
TargetDependency.remove(name: "Alamofire")
TargetDependency.thirdParty(name: "Alamofire")
TargetDependency.external(name: "Alamofire")
```
3. Extend the graph loading logic to load both, the local and the dependencies graph and then merge.

If we do it like this, the linting phase of project generation will run also for those third-party dependencies and catch potential integration issues. Moreover, users will be able to leverage the caching functionality of Tuist and thus, choose whether they want those third-party dependencies to be integrated as sources or binaries.

**Comment by fortmarek** [comment]
> Extend the graph loading logic to load both, the local and the dependencies graph and then merge.

I agree, but SPM nodes will still have to be built as `.xcframework`s, won't they? If so, this PR is still in the right direction imho as one way or another we'll need to build them. Eventually, we should:
1) Convert dependencies graph to use our primitives (in-memory) and build the frameworks using it (generate `.xcodeproj` in such a way that issues laid out by @thedavidharris are remedied if needed)
2) Save `ValueGraph` with `Codable` conformance where the nodes will be `.xcframework`s (or sources if user chooses that)
3) When generating project's graph, merge them with the one from dependencies

Or do you think @pepibumur we should take a different path for creating `.xcframework`s - or not creating `.xcframework`s at all? (but without xcframeworks I think we will lose a lot of benefits from `Dependencies.swift` SPM integration)

**Comment by laxmorek** [comment]
@pepibumur @fortmarek thank you for reviews. 

> I think we need to take a step back here

I think I can agree with you. 

> The workflow should be the following:
> 1. We resolve the dependencies using the SPM.
> 2. Serialize the graph of third-party dependencies into a disk as a JSON.
> 3. Extend the current graph model to support references to nodes that are in the third-party dependencies graph.
> 4. Implement the logic to merge the 2 graphs at loading time.

That is what I wanted to add in in upcoming PRs. However, it seems I have to change a direction a little and jump directly into the workflow you described. 

> I agree, but SPM nodes will still have to be built as .xcframeworks, won't they? If so, this PR is still in the right direction imho as one way or another we'll need to build them. Eventually, we should:

I'm not consider this PR as a wasted time. I learned a lot while I was working on it. I will keep local copy of this branch. I believe that changes I made here will be useful (and copy-paste'able) when I follow the new approach. 

=====================

Summarizing:
1. Im closing this PR.
2. I will clean up #2394 and prepare for merge since it solves `1. We resolve the dependencies using the SPM.`. 
3. I will continue my work on this feature.

**Comment by fortmarek** [comment]
We should probably still settle on how exactly we want to integrate SPM packages before continuing further, though, so the functionality done in this PR, with some adjustments, might still be necessary.

File: Sources/TuistSupport/XcodeBuild/XcodeBuildController.swift (hunk 1 above)
Comment (jesus-mg-ios) [reply Sources/TuistSupport/XcodeBuild/XcodeBuildController.swift:10]: Was it solved?

**Pull request closed without merging**